framework/src/dgkernels/ADDGKernel.C

```diff
@@ -0,0 +1,203 @@
+                (_sys.system().n_vars() * _sys.getMaxVarNDofsPerElem());
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    for (_i = 0; _i < test_space.size(); _i++)
+    {
+      DualReal residual = computeQpResidual(
+          (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? Moose::Element
+                                                                            : Moose::Neighbor);
+      for (_j = 0; _j < loc_phi.size(); _j++)
+        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * residual.derivatives()[ad_offset + _j];
+    }
+
+  accumulateTaggedLocalMatrix();
+
+  if (_has_diag_save_in && (type == Moose::ElementElement || type == Moose::NeighborNeighbor))
+  {
+    unsigned int rows = _local_ke.m();
+    DenseVector<Number> diag(rows);
+    for (unsigned int i = 0; i < rows; i++)
+      diag(i) = _local_ke(i, i);
+
+    Threads::spin_mutex::scoped_lock lock(_jacoby_vars_mutex);
+    for (const auto & var : _diag_save_in)
+    {
+      if (type == Moose::ElementElement)
+        var->sys().solution().add_vector(diag, var->dofIndices());
+      else
+        var->sys().solution().add_vector(diag, var->dofIndicesNeighbor());
+    }
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADDGKernel<compute_stage>::computeOffDiagJacobian(unsigned int jvar)
+{
+  DGKernelBase::computeOffDiagJacobian(jvar);
+}
+
+template <>
+void
+ADDGKernel<RESIDUAL>::computeOffDiagJacobian(unsigned int)
+{
+}
+
+template <ComputeStage compute_stage>
+void
+ADDGKernel<compute_stage>::computeOffDiagElemNeighJacobian(Moose::DGJacobianType type,
+                                                           unsigned int jvar)
+{
+  const VariableTestValue & test_space =
+      (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? _test : _test_neighbor;
+  const VariableTestValue & loc_phi =
+      (type == Moose::ElementElement || type == Moose::NeighborElement) ? _phi : _phi_neighbor;
+
+  if (type == Moose::ElementElement)
+    prepareMatrixTag(_assembly, _var.number(), jvar);
+  else
+    prepareMatrixTagNeighbor(_assembly, _var.number(), jvar, type);
+
+  size_t ad_offset = 0;
+  if (type == Moose::ElementElement || type == Moose::NeighborElement)
+    ad_offset = _var.number() * _sys.getMaxVarNDofsPerElem();
+  else
+    ad_offset = _var.number() * _sys.getMaxVarNDofsPerElem() +
+                (_sys.system().n_vars() * _sys.getMaxVarNDofsPerElem());
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    for (_i = 0; _i < test_space.size(); _i++)
+    {
+      DualReal residual = computeQpResidual(
+          (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? Moose::Element
+                                                                            : Moose::Neighbor);
+      for (_j = 0; _j < loc_phi.size(); _j++)
+        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * residual.derivatives()[ad_offset + _j];
+    }
+
+  accumulateTaggedLocalMatrix();
+}
+
+template class ADDGKernel<RESIDUAL>;
```

Comment:
You can use the macro that @dschwen created `adBaseClass(ADDGKernel)` which automatically does these two lines